electron/src/lib/AutomatedSingleSignOn.ts

```diff
@@ -0,0 +1,148 @@
+  public static readonly handleProtocolRequest = async (route: URL) => {
+    if (typeof route.pathname !== 'string') {
+      return;
+    }
+
+    const sendCodeToRenderer = () => {
+      const code = route.pathname.trim().substr(1);
+      main.webContents.send(EVENT_TYPE.ACCOUNT.SSO_LOGIN, code);
+    };
+
+    await app.whenReady();
+
+    const main = windowManager.getPrimaryWindow();
+    if (main.webContents.isLoading()) {
+      // App is booting
+      main.webContents.once('did-finish-load', () => sendCodeToRenderer());
+    } else {
+      if (!main.isVisible()) {
+        main.show();
+        main.focus();
+      }
+      sendCodeToRenderer();
+    }
+  };
+  private oldClipboard?: string;
+  private webview!: WebviewTag;
+  private webContents!: WebContents;
+
+  constructor(private readonly ssoCode: string) {}
+
+  private buildUrl() {
+    if (!this.webview.src) {
+      return 'about:blank';
+    }
+    const src = new URL(this.webview.src);
+    src.pathname = '/auth/';
+    src.hash = AutomatedSingleSignOn.SSO_LOGIN_HASH;
+    return src.href;
+  }
+
+  private async clickOn(element: string) {
+    await this.webContents.executeJavaScript(`document.querySelector('${element}').click();`);
+    await AutomatedSingleSignOn.sleep(AutomatedSingleSignOn.SSO_ACTION_DELAY);
+  }
+
+  private executeLogin() {
+    return new Promise((resolve, reject) => {
+      const failedToLoad = () => reject('Webview failed to load');
+      this.webview = document.querySelector('.Webview:not(.hide)') as WebviewTag;
+      this.webContents = this.webview.getWebContents();
+
+      this.webContents.once('did-fail-load', failedToLoad);
+      this.webContents.once('did-finish-load', async () => {
+        this.webContents.removeListener('did-fail-load', failedToLoad);
+        await this.clickOn('a[data-uie-name="go-sign-in-sso"]');
+        await this.clickOn('button[data-uie-name="do-sso-sign-in"]');
+        resolve();
+      });
+
+      this.webContents.loadURL(this.buildUrl());
+    });
+  }
+
+  private async onResponseReceived(event: CustomEvent) {
+    try {
+      if (event.detail && event.detail.reachedMaximumAccounts) {
+        throw new MaximumAccountsReachedError('Maximum account reached');
+      }
+
+      await this.executeLogin();
+    } catch (error) {
+      if (error instanceof MaximumAccountsReachedError) {
+        dialog.showMessageBox({
+          detail: getText('wrapperAddAccountErrorMessage'),
+          message: getText('wrapperAddAccountErrorTitle'),
+          type: 'warning',
+        });
+      }
+    } finally {
+      await this.restoreClipboard();
+    }
+  }
+
+  public async restoreClipboard() {
+    if (this.oldClipboard) {
+      await AutomatedSingleSignOn.clipboard.write(this.oldClipboard);
+    }
+  }
+
+  public async start() {
+    if (this.ssoCode.length > AutomatedSingleSignOn.SSO_CODE_MAX_LENGTH) {
+      return;
```

Comment:
We should show an error in the view that the code is invalid rather than doing nothing.

Reply:
This is not an attempt to validate the code, rather to prevent pasting too big data from being pasted to the clipboard.